Bug1431406 specific properties not removed from messages

gaohoward opened the pull request:
In a cluster if a node is shut down (or crashed) when a
message is being routed to a remote binding, a internal
property may be added to the message and persisted. The
name of the property is like _HQ_ROUTE_TOsf.my-cluster*.
if the node starts back, it will reroute this message
and if it goes to a local consumer, this property won't
get removed and goes to the client.

The fix is to remove this internal property before it
is sent to any client.

clebertsuconic commented:
@gaohoward this will break other cases... say you route a message towards a cluster.. with ROUTE_TO property.

now you crash a node.. and reload the messages.. the previous ROUTE_TO messages will need to be in place to send to that node.


a Message on a bridge belong to that node.. you can't remove these properties.. they are removed after the Bridge receives them.

gaohoward commented:
@clebertsuconic 
I see. I'll think of a solution and add a test.
Unfortunately it is in Artemis already. Can you revert the commit, the commit id is:
f465996444fdb6fcf33e1ce2262d6ea2aae0cdb1

Thanks
Howard

clebertsuconic commented:
@gaohoward I will revert it on artemis.

gaohoward commented:
I've some new findings. the internal property is sent to clients even there is no crash involved.
Consider this scenario with 2 nodes cluster:
node0 has q0 on address "address", node1 has q1 on the same address.
node0 has one local consumer to q0 and node1 has one local consumer to q1.
a producer sends a message to "address" at node0. 
each consumer will receive a copy of the message. But the consumer on q0/node0
gets the internal property (sth like _HQ_ROUTE_TOsf.cluster0.01699206-5752-11e7-baea-9fb1ef365e02), but consumer at node1 doesn't.

Same is true with Artemis.

Have to figure out a proper solution.

clebertsuconic commented:
remove it on the clientConsumer?

howardgao committed:
Bug1431406 specific properties not removed from messages

In a cluster if a message is routed to other nodes an internal
property (sth like _HQ_ROUTE_TOsfxxx) will be put to message
in order to route the message to its remote destination.
This property can be propogated to the consumers that
doesn't recognise the property.

For example if a 3rd party JMS consumer receives the message
it can get the property name by JMS api	Message.getPropertyNames()
and if the name is not confirm to some rules an erro will
occur.

The fix is that HorentQ message shouldn't expose such property
via JMS API.

gaohoward commented:
@clebertsuconic 
It is very hard to separate this property from core. Simple solution is just don't expose the property to JMS api.
The reason user can get this issue is that it uses jms.Message.getPropertyNames() for check each property name if it conforms certain naming convention. (like a dot in the name is not valid in WebLogic).
So simply don't return the property name to JMS getPropertyNames() should be ok.

Wdyt?
Howard

clebertsuconic commented:
Can you send this commit upstream as well?
